### ISSUE TITLE
Create Fusion Model Training Notebook

### ISSUE THREAD

**Issue by lukhsaankumar** [McMasterAI-Society/DeepFakeDetector#57]
## Issue Type
- [x] **Model**: ML model bug, training issue, or architecture problem
- [ ] **Data**: Dataset issue, preprocessing bug, or data pipeline problem  
- [ ] **Web**: Frontend bug or UI issue in the Next.js dashboard
- [ ] **API**: Backend API bug or FastAPI endpoint issue
- [ ] **Research**: Research question or experimental feature request
- [ ] **Documentation**: Documentation bug or improvement needed
- [ ] **Bug**: General bug fix needed
- [ ] **Enhancement**: New feature or improvement request

## Description

Create a Jupyter notebook that trains a fusion model on top of 3 submodels: CNN Transfer, Gradfield CNN, and **one** of either DeiT Distilled or ViT Base (user-selectable via a config variable at the top of the notebook). The fusion model learns to combine submodel probability outputs into a single `real/fake` prediction with a calibrated confidence score.

The notebook should be designed so that adding or removing submodels requires no architecture changes — the fusion model dynamically adapts to however many submodel outputs it receives. Swapping between DeiT and ViT should require changing a single variable only.

### Branch & Setup

Checkout the `fusion-notebook` branch and run the fetch script to pull all production model weights locally before starting:
```bash
git checkout fusion-notebook
python scripts/fetch_production_models.py
```

Model weights and configs will be available under the `models/` folder from the repo root:
```
models/
├── DeepFakeDetector--cnn-transfer/
│   ├── config.json
│   ├── preprocess.json
│   └── model.pth
├── DeepFakeDetector--deit-distilled/
│   ├── config.json
│   ├── preprocess.json
│   └── deit_distilled_custom_head.pt
├── DeepFakeDetector--vit-base/
│   ├── config.json
│   ├── preprocess.json
│   └── deepfake_vit_finetuned_wildfake.pth
└── DeepFakeDetector--gradfield-cnn/
    ├── config.json
    ├── preprocess.json
    └── gradient_field_cnn_v2.pth
```

### Submodel Architecture Reference

| Model | Weights File | Input Size | Preprocessing | Architecture |
|---|---|---|---|---|
| `cnn-transfer` | `model.pth` | 224×224 | ImageNet norm | EfficientNet-B0 |
| `deit-distilled` | `deit_distilled_custom_head.pt` | 224×224 | ImageNet norm | DeiT-B + Custom MLP head (768→512→2) |
| `vit-base` | `deepfake_vit_finetuned_wildfake.pth` | 224×224 | ImageNet norm | ViT-B/16 (timm) |
| `gradfield-cnn` | `gradient_field_cnn_v2.pth` | 256×256 | Sobel gradient + 0.5 norm | CompactGradientNet (5-channel input) |

> **Note on DeiT vs ViT:** Both use identical 224×224 ImageNet preprocessing and output a `prob_fake` float — the fusion layer sees no difference between them. The only implementation difference is that DeiT requires reconstructing its custom MLP head (768→512→2) before loading weights, while ViT is a clean `load_state_dict`. ViT is recommended as the default for simplicity.

---

## Deliverable

A complete, runnable Jupyter notebook with the following cell structure. **Both fusion models must be trained and uploaded to HuggingFace.** The meta-classifier is a stretch goal but should be attempted.

### Cell 1 — Configuration (User-Editable)

This is the **only cell the user needs to edit** to swap submodels or change settings. All downstream cells read from these variables.
```python
# ============================================================
# USER CONFIGURATION — edit this cell only
# ============================================================

# Choose which transformer submodel to include in the fusion.
# Options: "vit-base" | "deit-distilled"
TRANSFORMER_MODEL = "vit-base"

# Paths to model folders (relative to repo root)
MODEL_ROOT = "models"

# Active submodels — CNN and Gradfield are always included.
# The transformer slot is filled by TRANSFORMER_MODEL above.
ACTIVE_SUBMODELS = ["cnn-transfer", TRANSFORMER_MODEL, "gradfield-cnn"]

# Fusion settings
FUSION_THRESHOLD = 0.5
TEST_SPLIT = 0.2
RANDOM_SEED = 42

# HuggingFace org to upload to
HF_ORG = "DeepFakeDetector"
# ============================================================
```

All subsequent cells must reference `ACTIVE_SUBMODELS` and `TRANSFORMER_MODEL` — **no model names should be hardcoded anywhere else in the notebook.**

### Cell 2 — Setup & Imports

- Install/import all dependencies: `torch`, `timm`, `PIL`, `sklearn`, `numpy`, `joblib`, `huggingface_hub`
- Derive all paths from `MODEL_ROOT` and `ACTIVE_SUBMODELS`
- Print a summary of the active configuration so the user can confirm before running
```python
print(f"Active submodels: {ACTIVE_SUBMODELS}")
print(f"Transformer slot: {TRANSFORMER_MODEL}")
print(f"Fusion threshold: {FUSION_THRESHOLD}")
```

### Cell 3 — Define Preprocessing Functions
```python
def preprocess_standard(image: PIL.Image) -> torch.Tensor:
    """For cnn-transfer, deit-distilled, vit-base — 224×224, ImageNet norm"""

def preprocess_gradfield(image: PIL.Image) -> torch.Tensor:
    """For gradfield-cnn — 256×256, Sobel gradient, 5-channel input"""
```

### Cell 4 — Define Submodel Loaders

Define a loader for every available submodel. Each loader must:
1. Read `config.json` and `preprocess.json` from the model folder
2. Load the `.pth` weights using the appropriate architecture
3. Return a callable: `PIL.Image → {"prob_fake": float}`

| Model | Load Pattern |
|---|---|
| `cnn-transfer` | `torchvision.models.efficientnet_b0`, replace classifier, load `state_dict` |
| `vit-base` | `timm.create_model('vit_base_patch16_224')`, load `state_dict` directly |
| `deit-distilled` | `timm.create_model('deit_base_distilled_patch16_224')`, reconstruct `head`/`head_dist` as custom MLP (768→512→2), then load `state_dict` |
| `gradfield-cnn` | Custom `CompactGradientNet` class, Sobel preprocessing, 5-channel input |

All four loaders are always defined. Only the ones in `ACTIVE_SUBMODELS` are actually loaded and used:
```python
ALL_LOADERS = {
    "cnn-transfer":    load_cnn_transfer,
    "vit-base":        load_vit_base,
    "deit-distilled":  load_deit_distilled,
    "gradfield-cnn":   load_gradfield_cnn,
}

# Only instantiate active submodels
loaded_models = {name: ALL_LOADERS[name](f"{MODEL_ROOT}/DeepFakeDetector--{name}")
                 for name in ACTIVE_SUBMODELS}
```

### Cell 5 — Run Inference on Validation Set

- For each image in the validation set, run all models in `loaded_models` and collect `prob_fake`
- Store results as rows of `[probs..., true_label]` in the order defined by `ACTIVE_SUBMODELS`
- Save to `validation_data.json` for reproducibility — include `ACTIVE_SUBMODELS` as a field so the file is self-describing
```python
# validation_data.json schema
{
  "submodel_order": ["cnn-transfer", "vit-base", "gradfield-cnn"],
  "rows": [
    {"probs": [0.91, 0.74, 0.83], "label": 1},
    ...
  ]
}
```

### Cell 6 — Train Logistic Regression Fusion (Primary)
```python
from sklearn.linear_model import LogisticRegression

# X shape: (N, len(ACTIVE_SUBMODELS))
# y: true labels (0=real, 1=fake)
fusion_logreg = LogisticRegression(penalty='l2', C=1.0, random_state=RANDOM_SEED)
fusion_logreg.fit(X_train, y_train)

# Inspect per-submodel contribution
for name, weight in zip(ACTIVE_SUBMODELS, fusion_logreg.coef_[0]):
    print(f"  {name}: {weight:.4f}")
```

The logistic regression input dimension is `len(ACTIVE_SUBMODELS)` — swapping the transformer or adding a submodel requires only rerunning Cells 5 and 6, with no architecture changes.

### Cell 7 — Evaluate

Report the following metrics for any trained fusion model:
- AUROC
- Accuracy
- F1 Score
- Calibration Error

Save results to `test_metrics.json`. Include `submodel_order` and `transformer_model` as fields so results are traceable back to the exact configuration used.

### Cell 8 — Attempt Meta-Classifier Fusion (Stretch Goal)

If logistic regression AUROC < 0.90, or as an additional comparison, train a small neural meta-classifier. Input dimension must be derived from `ACTIVE_SUBMODELS`, not hardcoded:
```python
import torch.nn as nn

n = len(ACTIVE_SUBMODELS)  # dynamic — do not hardcode

fusion_meta = nn.Sequential(
    nn.Linear(n, 32),
    nn.ReLU(),
    nn.Dropout(0.3),
    nn.Linear(32, 1),
    nn.Sigmoid()
)
```

Compare AUROC against the logistic regression baseline and note which performs better.

### Cell 9 — Export & Upload to HuggingFace

Save and upload both models (or the meta-classifier if trained) to their respective HuggingFace repositories:

| Model | HuggingFace Repo |
|---|---|
| Logistic Regression Fusion | [`DeepFakeDetector/fusion-logreg`](https://huggingface.co/DeepFakeDetector/fusion-logreg) |
| Meta-Classifier Fusion | [`DeepFakeDetector/fusion-meta-classifier`](https://huggingface.co/DeepFakeDetector/fusion-meta-classifier) |

Each uploaded repo must contain:
```
├── config.json         # active submodel list, submodel_order, threshold, type
├── predict.py          # callable predict(submodel_outputs: dict) -> dict
├── fusion_logreg.pkl   # (logreg only) joblib dump
├── fusion_model.pt     # (meta-classifier only) torch state_dict
├── label_map.json      # {0: "real", 1: "fake"}
├── test_metrics.json   # AUROC, accuracy, F1, calibration error
└── README.md
```

**`config.json` schema** — `submodel_order` must reflect whichever transformer was active when trained:
```json
{
  "type": "probability_stacking_fusion",
  "name": "fusion-logreg",
  "version": "1.0.0",
  "submodels": [
    "DeepFakeDetector/cnn-transfer",
    "DeepFakeDetector/vit-base",
    "DeepFakeDetector/gradfield-cnn"
  ],
  "submodel_order": ["cnn-transfer", "vit-base", "gradfield-cnn"],
  "num_submodels": 3,
  "threshold": 0.5,
  "labels": {"0": "real", "1": "fake"}
}
```

**`predict.py` interface (both repos must match this exact signature):**
```python
def predict(submodel_outputs: dict, **kwargs) -> dict:
    """
    Args:
        submodel_outputs: {
            "cnn-transfer":  {"prob_fake": 0.92},
            "vit-base":      {"prob_fake": 0.74},
            "gradfield-cnn": {"prob_fake": 0.83},
        }
    Returns:
        {
            "pred": "fake",
            "pred_int": 1,
            "prob_fake": 0.91,
            "meta": { "type": "...", "submodel_probs": [...] }
        }
    """
```

### Acceptance Criteria

- [ ] Cell 1 contains a single `TRANSFORMER_MODEL` variable that switches between `"vit-base"` and `"deit-distilled"` with no other changes required
- [ ] All submodel loaders are defined for all 4 models, but only `ACTIVE_SUBMODELS` are instantiated
- [ ] No model names are hardcoded outside of Cell 1
- [ ] Validation set probabilities are collected and saved to `validation_data.json` including `submodel_order`
- [ ] Logistic regression fusion is trained with input dim derived from `len(ACTIVE_SUBMODELS)`
- [ ] Meta-classifier input dimension is also derived from `len(ACTIVE_SUBMODELS)`, not hardcoded
- [ ] AUROC is reported and saved to `test_metrics.json` for both fusion approaches
- [ ] Both models are uploaded to their respective HuggingFace repos with the required file structure
- [ ] `predict.py` in each repo matches the required function signature
- [ ] `config.json` in each repo reflects the actual `submodel_order` used during training

---

## Additional Context

- **Recommended default:** set `TRANSFORMER_MODEL = "vit-base"` — it loads cleanly with a direct `load_state_dict` and has no head reconstruction complexity
- **If using DeiT:** the `head` and `head_dist` layers must be replaced with the custom MLP (768→512→2) *before* calling `load_state_dict`, otherwise the weights will not map correctly
- Both DeiT and ViT use identical 224×224 ImageNet normalization preprocessing — the fusion layer sees no difference between them, only the loader changes
- The `gradfield-cnn` model uses a **non-standard 5-channel Sobel input** — its preprocessing must always be handled separately from the standard ImageNet pipeline
- The logistic regression fusion weights (`.coef_`) are a useful diagnostic for which submodels are contributing most — worth printing and including in the README
- The `predict.py` interface is standardised across fusion repos so the backend wrapper can swap between `fusion-logreg` and `fusion-meta-classifier` without any code changes
